### DISCUSSION TITLE
🐛 possible bug: can't migrate sqlite due to `FOREIGN KEY constraint failed`

### DISCUSSION THREAD

**Discussion by lukas-runge** [mikro-orm/mikro-orm#6295] [Q&A]
Hello everyone 👋,

the title already says it. I am using default migration configuration which should have foreign keys disabled by default, but this seems to not be the case. This is what my migration looks like:

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const { Migration } = require("@mikro-orm/migrations");

class Migration20241211232631_add_option_attributes extends Migration {
	async up() {
		// rename the old table
		this.addSql("alter table `attribute` rename to `temp_attribute`;");

		// create the new table with the new column and changed column type
		this.addSql(
			"create table `attribute` (`id` text not null, `name` text not null, `type` text not null, `default_value` text not null, `label` text null, `value_range` json null, `unit` json null, `options` json null, `inputs` json null, `outputs` json null, `mapping` json null, `product_id` text null, `instance_id` text null, constraint `attribute_product_id_foreign` foreign key(`product_id`) references `product`(`id`) on delete set null on update cascade, constraint `attribute_instance_id_foreign` foreign key(`instance_id`) references `product_instance`(`id`) on delete set null on update cascade, primary key (`id`));"
		);

		// copy the data from the old table to the new table
		this.addSql(
			"insert into `attribute` (`id`, `name`, `type`, `default_value`, `label`, `value_range`, `unit`, `options`, `inputs`, `outputs`, `mapping`, `product_id`, `instance_id`) select `id`, `name`, `type`, CAST(`default_value` AS TEXT), `label`, `value_range`, `unit`, null, `inputs`, `outputs`, `mapping`, `product_id`, `instance_id` from `temp_attribute`;"
		);

		// drop the old table
		this.addSql("drop table `temp_attribute`;");
		
		// create indices for faster querying
		this.addSql("create index `attribute_product_id_index` on `attribute` (`product_id`);");
		this.addSql("create index `attribute_instance_id_index` on `attribute` (`instance_id`);");
	}
}
exports.Migration20241211232631_add_option_attributes =
	Migration20241211232631_add_option_attributes;
```

And this is the error i get:
```bash
MigrationError: Migration Migration20241211232631_add_option_attributes (up) failed: Original error: drop table `temp_attribute`; - SQLITE_CONSTRAINT: FOREIGN KEY constraint failed
[...]
   {
    code: 'SQLITE_CONSTRAINT',
    errno: 19,
    sqlState: undefined,
    sqlMessage: undefined,
    errmsg: undefined
  },
  jse_cause: ForeignKeyConstraintViolationException: drop table `temp_attribute`; - SQLITE_CONSTRAINT: FOREIGN KEY constraint failed
  [...]
   {
    code: 'SQLITE_CONSTRAINT',
    errno: 19,
    sqlState: undefined,
    sqlMessage: undefined,
    errmsg: undefined
  },
  migration: {
    direction: 'up',
    name: 'Migration20241211232631_add_option_attributes',
    path: '/src/migrations/show/Migration20241211232631_add_option_attributes.js',
    context: {}
  }
}
```

Does anyone have an idea, what's going on here?

**Best regards**
@lukas-runge

**Comment by B4nan** [comment]
> using default migration configuration which should have foreign keys disabled by default

Why do you think the default is disabled FKs? That's not true at all.

> Does anyone have an idea, what's going on here?

No idea without a reproduction, you need to show what you are doing, not what you are getting generated.

**Reply by B4nan** [reply]
Or do you mean they are disabled when running the migrations? We use `pragma foreign_keys = off;` automatically, that's correct. Still, you need to show how to get to your error, not what it looks like from outside.

**Reply by B4nan** [reply]
Actually, #5993 changed the default for `disableForeignKeys`, its no longer enabled by default since v6.3.8

Maybe we should keep that enabled for sqlite because of their limitations of alter table queries.

**Reply by lukas-runge** [reply]
@B4nan thank you for your fast response! 🙏

> Maybe we should keep that enabled for sqlite because of their limitations of alter table queries.

Yeah - sqlites limitations for alter table queries are the whole reason, why I am doing all that "shenanigans". 😅


> Actually, #5993 changed the default for `disableForeignKeys`, its no longer enabled by default since v6.3.8

Good to know - this then also needs to then be changed in the documentation as well: https://mikro-orm.io/docs/migrations
There it still says, that the default for `disableForeignKeys` is `true`.

Sadly when setting `disableForeignKeys` to `true` this does still not resolve the issue when migrating. Here is my full config:

```typescript
showDb = await MikroORM.init({
	entities: SHOW_ENTITIES,
	driver: SqliteDriver,
	dbName: filePath,
	serialization: { forceObject: true },
	forceEntityConstructor: true,
	extensions: [Migrator],
	migrations: {
		path: `${dev ? "" : `${process.resourcesPath}/app.asar/`}src/migrations/show`,
		emit: "js",
		snapshotName: ".snapshot",
		disableForeignKeys: true
	},
	logger: isTest ? () => {} : (msg) => SystemStorageLogger.debug(msg)
});
```

**Reply by lukas-runge** [reply]
> Or do you mean they are disabled when running the migrations? We use `pragma foreign_keys = off;` automatically, that's correct. Still, you need to show how to get to your error, not what it looks like from outside.

This whole thread is only about running migrations. I haven't yet found the the time to build a reproduction. I'll do that as asap. This is why I started this out as a discussion first (also in hopes of me just doing something wrong and this not being a bug). 🙌

**Reply by lukas-runge** [reply]
I just wrapped the drop table command manually  in `PRAGMA foreign_keys = OFF;` to find a workaround for now. Sadly this does not resolve the error either. 😳 Executing the exact same SQL in DBeaver works as expected. 🤔

This is my migration code:
```typescript
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const { Migration } = require("@mikro-orm/migrations");

class Migration20241211232631_add_option_attributes extends Migration {
	async up() {
		// rename the old table
		this.addSql("alter table `attribute` rename to `temp_attribute`;");
		
		// create the new table with the new column and changed column type
		this.addSql(
			"create table `attribute` (`id` text not null, `name` text not null, `type` text not null, `default_value` text not null, `label` text null, `value_range` json null, `unit` json null, `options` json null, `inputs` json null, `outputs` json null, `mapping` json null, `product_id` text null, `instance_id` text null, constraint `attribute_product_id_foreign` foreign key(`product_id`) references `product`(`id`) on delete set null on update cascade, constraint `attribute_instance_id_foreign` foreign key(`instance_id`) references `product_instance`(`id`) on delete set null on update cascade, primary key (`id`));"
		);
		
		// copy the data from the old table to the new table
		this.addSql(
			"insert into `attribute` (`id`, `name`, `type`, `default_value`, `label`, `value_range`, `unit`, `options`, `inputs`, `outputs`, `mapping`, `product_id`, `instance_id`) select `id`, `name`, `type`, CAST(`default_value` AS TEXT), `label`, `value_range`, `unit`, null, `inputs`, `outputs`, `mapping`, `product_id`, `instance_id` from `temp_attribute`;"
		);
		
		// drop the old table
		this.addSql('PRAGMA foreign_keys = OFF;');
		this.addSql("drop table `temp_attribute`;");
		this.addSql('PRAGMA foreign_keys = ON;');
		
		// create indices for faster querying
		this.addSql("create index `attribute_product_id_index` on `attribute` (`product_id`);");
		this.addSql("create index `attribute_instance_id_index` on `attribute` (`instance_id`);");
	}
}
exports.Migration20241211232631_add_option_attributes =
	Migration20241211232631_add_option_attributes;
```

and this is the SQL I successfully validated in DBeaver:
```SQL
BEGIN TRANSACTION;

alter table `attribute` rename to `temp_attribute`;

create table `attribute` (`id` text not null, `name` text not null, `type` text not null, `default_value` text not null, `label` text null, `value_range` json null, `unit` json null, `options` json null, `inputs` json null, `outputs` json null, `mapping` json null, `product_id` text null, `instance_id` text null, constraint `attribute_product_id_foreign` foreign key(`product_id`) references `product`(`id`) on delete set null on update cascade, constraint `attribute_instance_id_foreign` foreign key(`instance_id`) references `product_instance`(`id`) on delete set null on update cascade, primary key (`id`));
insert into `attribute` (`id`, `name`, `type`, `default_value`, `label`, `value_range`, `unit`, `options`, `inputs`, `outputs`, `mapping`, `product_id`, `instance_id`) select `id`, `name`, `type`, CAST(`default_value` AS TEXT), `label`, `value_range`, `unit`, null, `inputs`, `outputs`, `mapping`, `product_id`, `instance_id` from `temp_attribute`;

PRAGMA foreign_keys = OFF;
drop table `temp_attribute`;
PRAGMA foreign_keys = ON;
		
create index `attribute_product_id_index` on `attribute` (`product_id`);
create index `attribute_instance_id_index` on `attribute` (`instance_id`);

COMMIT;
```

**Reply by B4nan** [reply]
I read somewhere the pragma statement needs to be executed separately and IIRC we batch all the queries into one, maybe its about that. Or the opposite 🙃

**Reply by lukas-runge** [reply]
> I read somewhere the pragma statement needs to be executed separately and IIRC we batch all the queries into one, maybe its about that. Or the opposite 🙃

Yeah that could be the reason, why it does not work. I tried working around that by committing the transaction and beginning a new one, but that results in a new error. How could I work around that?

```javascript
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const { Migration } = require("@mikro-orm/migrations");

class Migration20241211232631_add_option_attributes extends Migration {
	async up() {
		this.addSql("COMMIT;");
		this.addSql("PRAGMA foreign_keys = OFF;");
		this.addSql("BEGIN TRANSACTION;");

		// rename the old table
		this.addSql(
			"alter table `attribute` rename to `temp_attribute`;"
		);

		// create the new table with the new column and changed column type
		this.addSql(
			"create table `attribute` (`id` text not null, `name` text not null, `type` text not null, `default_value` text not null, `label` text null, `value_range` json null, `unit` json null, `options` json null, `inputs` json null, `outputs` json null, `mapping` json null, `product_id` text null, `instance_id` text null, constraint `attribute_product_id_foreign` foreign key(`product_id`) references `product`(`id`) on delete set null on update cascade, constraint `attribute_instance_id_foreign` foreign key(`instance_id`) references `product_instance`(`id`) on delete set null on update cascade, primary key (`id`));"
		);

		// copy the data from the old table to the new table
		this.addSql(
			"insert into `attribute` (`id`, `name`, `type`, `default_value`, `label`, `value_range`, `unit`, `options`, `inputs`, `outputs`, `mapping`, `product_id`, `instance_id`) select `id`, `name`, `type`, CAST(`default_value` AS TEXT), `label`, `value_range`, `unit`, null, `inputs`, `outputs`, `mapping`, `product_id`, `instance_id` from `temp_attribute`;"
		);

		// drop the old table
		this.addSql("drop table `temp_attribute`;");

		// create indices for faster querying
		this.addSql("create index `attribute_product_id_index` on `attribute` (`product_id`);");
		this.addSql("create index `attribute_instance_id_index` on `attribute` (`instance_id`);");

		this.addSql("PRAGMA foreign_keys = ON;");
	}
}
exports.Migration20241211232631_add_option_attributes =
	Migration20241211232631_add_option_attributes;
```

The error:
```bash
MigrationError: Migration Migration20241211232631_add_option_attributes (up) failed: Original error: RELEASE SAVEPOINT trx9; - SQLITE_ERROR: no such savepoint: trx9
    at /node_modules/umzug/lib/umzug.js:169:27
    at async Umzug.runCommand (/node_modules/umzug/lib/umzug.js:125:20)
    at async Migrator.runInTransaction (/node_modules/@mikro-orm/migrations/Migrator.js:313:21)
    at async SqliteConnection.transactional (/node_modules/@mikro-orm/knex/AbstractSqlConnection.js:58:25)
    at async Module.openShow (/src/lib/server/db/show.server.ts:85:4)
    at async POST (/src/routes/api/+server.ts:12:4)
    at async Module.render_endpoint (/node_modules/@sveltejs/kit/src/runtime/server/endpoint.js:43:18)
    at async resolve (/node_modules/@sveltejs/kit/src/runtime/server/respond.js:443:17)
    at async eval (/src/hooks.server.ts:43:15)
    at async Object.handle (/src/hooks.server.ts:18:9) {
  cause: [Error: RELEASE SAVEPOINT trx9; - SQLITE_ERROR: no such savepoint: trx9] {
    errno: 1,
    code: 'SQLITE_ERROR'
  },
  jse_cause: [Error: RELEASE SAVEPOINT trx9; - SQLITE_ERROR: no such savepoint: trx9] {
    errno: 1,
    code: 'SQLITE_ERROR'
  },
  migration: {
    direction: 'up',
    name: 'Migration20241211232631_add_option_attributes',
    path: '/src/migrations/show/Migration20241211232631_add_option_attributes.js',
    context: {}
  }
}
```

**Reply by B4nan** [reply]
All migrations are executed inside a master transaction, the separate migrations are nested transactions, so savepoints. You can't just do this inside a raw query most likely.

I'd say you should sort the FKs out instead of trying to ignore them. I'd say you have some FKs targeting your old table.

**Reply by lukas-runge** [reply]
> All migrations are executed inside a master transaction, the separate migrations are nested transactions, so savepoints. You can't just do this inside a raw query most likely.
> 
> I'd say you should sort the FKs out instead of trying to ignore them. I'd say you have some FKs targeting your old table.

There are no FKs targeting my old table (but even if they were this should work as a table with the same name and same data will be in place after the migration). As far as I can see SQLite does not provide any way of dropping foreign constraints without using `PRAGMA foreign_keys = OFF;`. Do you know of any other way to, as you said, "sort out" the FKs? 🙌

**Reply by B4nan** [reply]
I guess you did the migration by hand? Have you tried using the schema generator?

**Reply by lukas-runge** [reply]
> I guess you did the migration by hand? Have you tried using the schema generator?

I started off with the schema generator, sadly it didn't generate any migration for the changed column type (Integer > Text). This is why I had to build that on my own. 😑

**Reply by lukas-runge** [reply]
I now worked around it by disabling `allOrNothing` as well as `transactional`. This also resulted in the `disableForeignKeys` switch working again. It seems like the pragma statement is send within the context of the transaction which renders it useless as you already pointed out above. The moment that there is no transaction wrapped around it anymore the `disableForeignKeys` flag works as intended. But now, as there is no transaction in place anymore that would save me from a failing migration, I now programmatically create a backup copy of the db before migration and roll back to that when there is any exception thrown while migrating.

**Reply by B4nan** [reply]
Can you share the entities before and after, so I can test all of this (failing to detect the schema difference, as well as the FK issue)? This is surely something to fix, I wonder how come nobody reported this before.

**Reply by lukas-runge** [reply]
> Can you share the entities before and after, so I can test all of this (failing to detect the schema difference, as well as the FK issue)? This is surely something to fix, I wonder how come nobody reported this before.

@B4nan Just created a reproduction repo for you at https://github.com/lukas-runge/mikroorm-sqlite-migration-reproduction
Also as this is a confirmed bug now, I created an issue for it here: https://github.com/mikro-orm/mikro-orm/issues/6300 🙌

**Reply by B4nan** [reply]
Well, I asked for the reproduction so I can confirm it, so don't rush the resolution 🙃

**Reply by lukas-runge** [reply]
> Well, I asked for the reproduction so I can confirm it, so don't rush the resolution 🙃 

I didn't mean to bother you. No worries. Take your time. I am happy with my workaround for now. 🙌
